### ISSUE TITLE
[Bug]: DFlash draft lm_head 被覆盖导致接受率从 25% 跌至 1.7%

### ISSUE THREAD

**Issue by nevin001** [vllm-project/vllm-ascend#11374]
### Your current environment

[env.txt](https://github.com/user-attachments/files/29618651/env.txt)

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of above commands here
```

</details>

## 问题描述

使用 RedHat 风格的 DFlash 推测解码草稿模型（如 `RedHatAI/Qwen3-8B-speculator.dflash` ）与 Qwen3-8B 目标模型组合时，在 vllm-ascend 上 speculative decoding 的接受率从正常的 **~25%**骤降到 **~1.7%**，几乎每个推测 token 都被拒绝。

根因是 vllm-ascend 的 `AscendSpecDecodeBaseProposer._maybe_share_lm_head` 强制把 draft 模型的 `lm_head` 替换为 target 模型的 `lm_head`，而 RedHat 风格的 DFlash 草稿训练时使用了一个**专门的 32000 行 lm_head**，被替换后 logits 映射到错误的词表，verifier 几乎全部拒绝。

这是一个 **已修复 bug 的回归问题**——`PR #9795 (commit 4ed6eb07)` 已经修复过这个问题，但 `PR #9946 (commit 367b8e62)` 的"消除所有 patch 行为"重构把修复覆盖了，至今未恢复。

## 复现步骤

1. 准备一个 RedHat 风格的 DFlash 草稿模型（`draft_vocab_size=32000`），例如：
   - 直接使用 `RedHatAI/Qwen3-8B-speculator.dflash`
2. 准备目标模型 `Qwen3-8B`
3. 启动 vllm-ascend，启用 dflash 推测解码：

   ```bash
   vllm serve /data/Qwen3-8B \
     --speculative-model /path/to/dflash-draft \
     --speculative-method dflash \
     --num-speculative-tokens 7 \
     --port 8401
   ```

4. 用 vllm bench 跑推理：

   ```bash
   vllm bench serve \
     --model /path/to/dflash-draft \
     --tokenizer /data/Qwen3-8B \
     --dataset-name hf \
     --dataset-path /data/wangchunning/datasets/likaixin/InstructCoder \
     --hf-name likaixin/InstructCoder \
     --hf-split train \
     --num-prompts 100 \
     --temperature 0
   ```

5. 观察结果：
   - **预期**：mean accept length ≈ 3.3（HumanEval）/ 3.8（GSM8K），整体接受率约 25%
   - **实际**：mean accept length ≈ 1.02，整体接受率约 1.7%

6. 同步观察 vllm-ascend 启动日志，应能看到：
   ```
   INFO ... Loading EAGLE or DFLASH LM head weights from the target model.
   ```
   这行日志是 bug 的明确标识——正常情况下 dflash 不应输出这行。

## 影响

- **影响范围**：所有使用 DFlash 推测解码 +自带lm_head的场景。
- **典型场景**：使用 RedHat 发布的 DFlash 草稿模型。
- **现象**：
  - 接受率从约 25% 跌至 1.7%（下降 15 倍）
  - 草稿完全失效，等价于"无推测解码"，TTFT 与生成时延明显恶化
  - vLLM 日志可见一行：`Loading EAGLE or DFLASH LM head weights from the target model.`，是问题的明确标识

## 环境信息

- vLLM 版本：v0.20.2
- vllm-ascend 版本：v0.20.2rc
- 目标模型：Qwen3-8B
- 草稿模型：`RedHatAI/Qwen3-8B-speculator.dflash`
- NPU：Ascend 910B4-1

**Comment by MengqingCao** [comment]
@lilinsiman please take a look at the precision issue on DFlash.

**Comment by crystalww** [comment]
Any updates？ vllm-0.23.0 has the same problem.
```
vllm serve /dev/shm/Qwen3.6-35B-A3B \
--trust-remote-code \
--enable-prompt-tokens-details \
--tensor-parallel-size 4 \
--max-num-seqs 128 \
--max-num-batched-tokens=8K \
--reasoning-parser qwen3 \
--enable-auto-tool-choice \
--tool-call-parser qwen3_xml \
--enable-prefix-caching --mamba-cache-mode align \
--enable-expert-parallel \
--override-generation-config '{"temperature":0.6}' \
--served-model-name qwen \
--max-model-len 262144 \
--enable-mfu-metrics \
--speculative-config '{"num_speculative_tokens":8,"method":"dflash","model":"/dev/shm/Qwen3.6-35B-A3B-DFlash"}' \
--additional-config '{"ascend_compilation_config":{"fuse_allreduce_rms":true},"multistream_overlap_shared_expert":true,"multistream_overlap_gate":true,"enable_matmul_allreduce":true}' \
--compilation-config '{"cudagraph_mode":"FULL_DECODE_ONLY"}' \
--host :: --port $PORT0
```
<img width="3182" height="1180" alt="Image" src="https://github.com/user-attachments/assets/83061418-bebd-4f80-aa2f-f6fd2a3bfc98" />
